packages/docs/content/v4/index.mdx

```diff
@@ -0,0 +1,903 @@
+```
+
+Then to run a particular benchmark:
+
+```sh
+$ pnpm bench <name>
+```
+
+### 2.6x faster string parsing
+
+```sh
+$ pnpm bench string
+runtime: node v22.13.0 (arm64-darwin)
+
+benchmark      time (avg)             (min … max)       p75       p99      p999
+------------------------------------------------- -----------------------------
+• z.string().parse
+------------------------------------------------- -----------------------------
+zod3          348 µs/iter       (299 µs … 743 µs)    362 µs    494 µs    634 µs
+zod4          132 µs/iter       (108 µs … 348 µs)    162 µs    269 µs    322 µs
+
+summary for z.string().parse
+  zod4
+   2.63x faster than zod3
+```
+
+### 3x faster array parsing
+
+```sh
+$ pnpm bench array
+runtime: node v22.13.0 (arm64-darwin)
+
+benchmark      time (avg)             (min … max)       p75       p99      p999
+------------------------------------------------- -----------------------------
+• z.array() parsing
+------------------------------------------------- -----------------------------
+zod3          162 µs/iter       (141 µs … 753 µs)    152 µs    291 µs    513 µs
+zod4       54'282 ns/iter    (47'084 ns … 669 µs) 50'833 ns    185 µs    233 µs
+
+summary for z.array() parsing
+  zod4
+   2.98x faster than zod3
+```
+
+### 7x faster object parsing
+
+This runs the [Moltar validation library benchmark](https://moltar.github.io/typescript-runtime-type-benchmarks/). 
+
+```sh
+$ pnpm bench object-moltar
+benchmark      time (avg)             (min … max)       p75       p99      p999
+------------------------------------------------- -----------------------------
+• z.object() safeParse
+------------------------------------------------- -----------------------------
+zod3          767 µs/iter     (735 µs … 3'136 µs)    775 µs    898 µs  3'136 µs
+zod4          110 µs/iter     (102 µs … 1'291 µs)    105 µs    217 µs    566 µs
+
+summary for z.object() safeParse
+  zod4
+   6.98x faster than zod3
+```
+
+## 20x reduction in `tsc` instantiations
+
+Consider the following simple file:
+
+```ts
+import * as z from "zod";
+
+export const A = z.object({
+  a: z.string(),
+  b: z.string(),
+  c: z.string(),
+  d: z.string(),
+  e: z.string(),
+});
+
+export const B = A.extend({
+  f: z.string(),
+  g: z.string(),
+  h: z.string(),
+});
+```
+
+Compiling this file with `tsc --extendedDiagnostics` using `zod@3` results in >25000 type instantiations. With `zod@4` it only results in ~1100. 
+
+<Callout>
+
+The Zod repo contains a `tsc` benchmarking playground. Try this for yourself using the compiler benchmarks in `packages/tsc`. The exact numbers may change as the implementation evolves. 
+
+```sh
+$ cd packages/tsc
+$ pnpm bench object-with-extend
+```
+</Callout>
+
+
+More importantly, Zod 4 has redesigned and simplified the generics of `ZodObject` and other schema classes to avoid some pernicioius "instantiation explosions". The following script contains a number of chained called to `.extend()` and `.omit()`.
```

Comment:
```suggestion
More importantly, Zod 4 has redesigned and simplified the generics of `ZodObject` and other schema classes to avoid some pernicioius "instantiation explosions". The following script contains a number of chained calls to `.extend()` and `.omit()`.
```